# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Hania-Haris/FlyRank-AI/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

**Provisional lane: Lane 1 — Ranking Signal Analysis**

I want to investigate which observable content and search signals are associated with differences in search visibility, clicks, engagement, and page movement. I chose this lane because the starter dataset contains multiple measurable signals about search performance and content characteristics, which gives me an opportunity to explore relationships before deciding whether a more complex model is useful. I am particularly interested in identifying which signals appear consistently useful and whether their relationships differ across levels of search performance. This is a provisional choice, so I may narrow or change the question as I learn more from the data during the next few weeks.

In [24]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*


**Research question:** Which observable content and search signals are associated with differences in search visibility, clicks, engagement, and page movement?

**Decision:** The decision is which measurable signals and page characteristics should receive more attention when investigating changes in search performance.

**Who acts:** An SEO or content analyst would use the findings to decide which signals to investigate first when diagnosing changes in visibility, clicks, CTR, or engagement.

**Action:** Based on the eventual findings, the analyst could investigate pages or groups of pages showing unusual patterns, compare them with similar pages, and decide whether further content, metadata, technical, or monitoring work is warranted.

**Output:** The project will eventually produce a signal report containing measured relationships, grouped comparisons, charts, and practical recommendations about which signals appear useful for further investigation.

**Cost of a wrong call:** A false interpretation could cause an analyst to spend time investigating an unimportant signal or overlook a signal that deserves attention. It could also lead to an incorrect recommendation if correlation is mistaken for causation. Therefore, the findings will be treated as decision support rather than guaranteed explanations of search performance.

**Why data or ML may help:** The dataset contains many observable signals related to search performance, content characteristics, freshness, visibility, and engagement. Some relationships may be difficult to identify from individual examples or simple rules. Exploratory analysis can reveal useful patterns, while statistical or machine-learning methods may help quantify and compare multiple signals if they add value beyond simple summaries. A complex model will only be justified if it improves the decision or understanding enough to warrant its additional complexity.


In [25]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [26]:
import os
import subprocess

repo_url = "https://github.com/Hania-Haris/FlyRank-AI"
repo_path = "/content/FlyRank-AI"

if not os.path.exists(repo_path):
    subprocess.run(
        ["git", "clone", repo_url, repo_path],
        check=True
    )

os.chdir(repo_path)

print("Working directory:", os.getcwd())

Working directory: /content/FlyRank-AI


In [27]:
import pandas as pd
import numpy as np

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Rows:", len(df))
print("Columns:", len(df.columns))

Rows: 30000
Columns: 44


In [28]:
declining_rate = (
    df["trend_direction"]
    .str.lower()
    .eq("down")
    .mean() * 100
)

print(f"Pages marked as declining: {declining_rate:.2f}%")

Pages marked as declining: 54.21%


In [29]:
median_impressions = df["impressions_90d"].median()

print(f"Median 90-day impressions: {median_impressions:.1f}")

Median 90-day impressions: 731.0


In [30]:
median_ctr = df["ctr"].median()

print(f"Median CTR: {median_ctr:.3f}%")

Median CTR: 0.070%


In [31]:
# Relationship between impressions and clicks
impressions_clicks_corr = (
    df[["impressions_90d", "clicks_90d"]]
    .corr()
    .loc["impressions_90d", "clicks_90d"]
)

# Relationship between content age and impressions
age_impressions_corr = (
    df[["content_age_days", "impressions_90d"]]
    .corr()
    .loc["content_age_days", "impressions_90d"]
)

# Relationship between average position and CTR
# Zero position means no position data, so exclude those rows.
position_ctr = df[
    ["avg_position", "ctr"]
].copy()

position_ctr = position_ctr[
    position_ctr["avg_position"] > 0
]

position_ctr_corr = (
    position_ctr
    .corr()
    .loc["avg_position", "ctr"]
)

print(f"Impressions vs clicks correlation: {impressions_clicks_corr:.3f}")
print(f"Content age vs impressions correlation: {age_impressions_corr:.3f}")
print(f"Average position vs CTR correlation: {position_ctr_corr:.3f}")

Impressions vs clicks correlation: 0.696
Content age vs impressions correlation: -0.001
Average position vs CTR correlation: -0.080


54.21% declining

731 median impressions

0.070% median CTR

0.696 impressions-clicks correlation

-0.001 age-impressions correlation

-0.080 position-CTR correlation

### What the starter data suggests

The starter dataset contains 30,000 content items with a range of observable search and content signals. In my initial check, **54.21%** of pages were marked as declining. The median 90-day impression count was **731**, while the median CTR was **0.070%**.

The initial association checks also produced some interesting differences. Impressions and clicks had a correlation of **0.696**, indicating a relatively strong positive association in this dataset. In contrast, content age and impressions had a correlation of only **-0.001**, while average position and CTR had a correlation of **-0.080** after excluding rows without position data.

These results make Lane 1 worth investigating because the dataset contains substantial variation in search performance and the relationships between signals are not all equally strong. In particular, the difference between the stronger impressions-clicks association and the much weaker associations involving content age and average position suggests that the relationships between search signals deserve more detailed investigation rather than assuming every available signal is equally useful.

### Initial interpretation

The correlation results are only a starting point. The relatively strong impressions-clicks association is expected to be worth examining further, while the very weak correlations for content age versus impressions and average position versus CTR suggest that simple linear relationships may not fully describe these signals. I will investigate whether grouped comparisons, non-linear relationships, or combinations of signals provide additional useful evidence.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*


At this stage, I can identify and measure associations between observable signals in the starter dataset. I can report correlations, grouped differences, and descriptive patterns, and I can investigate whether simple analytical or machine-learning methods provide additional evidence.

I cannot claim that a signal is a confirmed Google ranking factor, that one signal causes another to change, or that changing a particular content characteristic will definitely improve search performance. The available data is observational rather than the result of a controlled experiment.

For example, the correlation of 0.696 between impressions and clicks does not mean that increasing impressions will necessarily cause clicks to increase. Similarly, the near-zero correlation between content age and impressions does not prove that content age has no relationship with search performance; it only describes the simple linear association observed in this dataset.

I will therefore use careful language such as "associated with," "correlated with," "higher or lower on average," and "observed in this dataset." Any eventual recommendation will be treated as decision support rather than causal proof.

In [32]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("Careful-claims check:")
print("✓ Findings are observational")
print("✓ Correlation is not treated as causation")
print("✓ No claims about Google's ranking algorithm")
print("✓ Results will be treated as decision support")


Careful-claims check:
✓ Findings are observational
✓ Correlation is not treated as causation
✓ No claims about Google's ranking algorithm
✓ Results will be treated as decision support


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.